spark/src/main/scala/io/delta/sharing/spark/filters/JsonPredicates.scala

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.sharing.spark.filters
+
+import scala.collection.mutable.ListBuffer
+
+import com.fasterxml.jackson.annotation.{JsonSubTypes, JsonTypeInfo}
+
+/**
+ * The evaluation context in which operations will be evaluated.
+ *
+ * @param partitionValues The partition values associated with parquet files.
+ */
+case class EvalContext(partitionValues: Map[String, String])
+
+/**
+ * The data types supported by the filtering operations.
+ */
+object OpDataTypes {
+  val BoolType = "bool"
+  val IntType = "int"
+  val LongType = "long"
+  val StringType = "string"
+  val DateType = "date"
+
+  val supportedTypes = Set(BoolType, IntType, LongType, StringType, DateType)
+}
+
+/**
+ * The base operation and the associated json type and sub type tagging needed
+ * to convert case class hierarchy into json.
+ *
+ * If we add support for a new operation, we need to add an entry below
+ * which will tag the operation with the given name.
+ */
+@JsonTypeInfo(
+    use = JsonTypeInfo.Id.NAME,
+    include = JsonTypeInfo.As.PROPERTY,
+    property = "op")
+@JsonSubTypes(
+  Array(
+    new JsonSubTypes.Type(value = classOf[ColumnOp], name = "column"),
+    new JsonSubTypes.Type(value = classOf[LiteralOp], name = "literal"),
+    new JsonSubTypes.Type(value = classOf[IsNullOp], name = "isNull"),
+    new JsonSubTypes.Type(value = classOf[EqualOp], name = "equal"),
+    new JsonSubTypes.Type(value = classOf[LessThanOp], name = "lessThan"),
+    new JsonSubTypes.Type(value = classOf[LessThanOrEqualOp], name = "lessThanOrEqual"),
+    new JsonSubTypes.Type(value = classOf[GreaterThanOp], name = "greaterThan"),
+    new JsonSubTypes.Type(value = classOf[GreaterThanOrEqualOp], name = "greaterThanOrEqual"),
+    new JsonSubTypes.Type(value = classOf[AndOp], name = "and"),
+    new JsonSubTypes.Type(value = classOf[OrOp], name = "or"),
+    new JsonSubTypes.Type(value = classOf[NotOp], name = "not")
+  )
+)
+trait BaseOp {
+
+  // Performs validations on this operation.
+  //
+  // The method throws an exception if any of the validations fail.
+  @throws[IllegalArgumentException]
+  def validate(): Unit
+
+  // Evaluates this operation in the given context.
+  //
+  // If the method throws an exception, the caller should skip filtering for this context.
+  @throws[IllegalArgumentException]
+  def eval(ctx: EvalContext): Any
```

Comment:
I am uncomfortable that `Any` can be drastically different types:
- (String, String)
- Boolean

Can the predicate evaluation return something better than just `Boolean`?